rclrs/src/node/graph.rs

```diff
@@ -0,0 +1,568 @@
+            (
+                slice::from_raw_parts(rcl_names.data, rcl_names.size),
+                slice::from_raw_parts(rcl_namespaces.data, rcl_namespaces.size),
+                slice::from_raw_parts(rcl_enclaves.data, rcl_enclaves.size),
+            )
+        };
+
+        let zipped_names = names_slice
+            .iter()
+            .zip(namespaces_slice.iter())
+            .zip(enclaves_slice.iter())
+            .map(|((name, namespace), enclave)| unsafe {
+                (
+                    CStr::from_ptr(*name).to_string_lossy().into_owned(),
+                    CStr::from_ptr(*namespace).to_string_lossy().into_owned(),
+                    CStr::from_ptr(*enclave).to_string_lossy().into_owned(),
+                )
+            })
+            .collect();
+
+        Ok(zipped_names)
+    }
+
+    /// Count the number of publishers for a given topic
+    pub fn count_publishers(&self, topic: &str) -> Result<usize, RclrsError> {
+        let topic_name = CString::new(topic).unwrap();
+        let mut count: usize = 0;
+
+        // SAFETY: No preconditions for this function
+        unsafe {
+            rcl_count_publishers(&*self.rcl_node_mtx.lock(), topic_name.as_ptr(), &mut count)
+                .ok()?
+        };
+        Ok(count)
+    }
+
+    /// Count the number of subscribers for a given topic
+    pub fn count_subscribers(&self, topic: &str) -> Result<usize, RclrsError> {
+        let topic_name = CString::new(topic).unwrap();
+        let mut count: usize = 0;
+
+        // SAFETY: No preconditions for this function
+        unsafe {
+            rcl_count_subscribers(&*self.rcl_node_mtx.lock(), topic_name.as_ptr(), &mut count)
+                .ok()?
+        };
+        Ok(count)
+    }
+
+    /// Get topic publisher info
+    pub fn get_publishers_info_by_topic(
+        &self,
+        topic: &str,
+    ) -> Result<Vec<TopicEndpointInfo>, RclrsError> {
+        self.get_publisher_subscriber_info_by_topic(topic, rcl_get_publishers_info_by_topic)
+    }
+
+    /// Get topic subscriber info
+    pub fn get_subscribers_info_by_topic(
+        &self,
+        topic: &str,
+    ) -> Result<Vec<TopicEndpointInfo>, RclrsError> {
+        self.get_publisher_subscriber_info_by_topic(topic, rcl_get_subscriptions_info_by_topic)
+    }
+
+    /// Get an rcl names_and_types function, with a "demangle" argument
+    fn get_names_and_types_by_node_with_demangle_arg(
+        &self,
+        node: &str,
+        namespace: &str,
+        getter: unsafe extern "C" fn(
+            *const rcl_node_t,
+            *mut rcl_allocator_t,
+            bool,
+            *const ::std::os::raw::c_char,
+            *const ::std::os::raw::c_char,
+            *mut rcl_names_and_types_t,
+        ) -> rcl_ret_t,
+    ) -> Result<TopicNamesAndTypes, RclrsError> {
+        let mut rcl_names_and_types = unsafe { rmw_get_zero_initialized_names_and_types() };
+        let node_name = CString::new(node).unwrap();
```

Comment:
For conversion to C string types, we're using this error handling approach: https://github.com/ros2-rust/ros2_rust/blob/3194b772762c5ba410e97d838930fbd690e7b0db/rclrs/src/node/builder.rs#L236